### DISCUSSION TITLE
hello_world.py - no output and (infinite?) loop

### DISCUSSION THREAD

**Discussion by elim17** [USC-ACTLab/crazyswarm#328] [Q&A]
Hello,

We are having trouble with getting the crazyfly(s) to fly. I believe we have everything set up correctly. 

- Using Motive Optitrack for motion capture on its own workstation(running windows)
- On a separate linux workstation, the first step is to run this base.launch file(below). It opens ros and sets up a vrpn client node. Note: the machine name with tritons at the Botton of the file is for another project also using the motion capture system. 
 ```<launch>
    <!-- include arguments at top -->
    <arg name="optitrack_server" default="192.168.1.69" />

    <!-- <node pkg="mats" type="status_tritons" name="status_tritons" respawn="true" /> -->

    <!-- below launches the VRPN client -->
    <!-- borrowed from sample.launch -->
    <node pkg="vrpn_client_ros" type="vrpn_client_node" name="vrpn_client_node">
        <rosparam subst_value="true">
            server: $(arg optitrack_server)
            port: 3883
            update_frequency: 100.0
            frame_id: world
            use_server_time: false
            broadcast_tf: true
            refresh_tracker_frequency: 1.0
        </rosparam>
    </node>

    <node pkg="mats" type="clock_server.py" name="clock_server" />

    <machine name="triton01" address="192.168.1.101" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
    <machine name="triton02" address="192.168.1.102" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
    <machine name="triton03" address="192.168.1.103" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
    <machine name="triton04" address="192.168.1.104" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
    <machine name="triton05" address="192.168.1.105" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
    <machine name="triton06" address="192.168.1.106" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
    <machine name="triton07" address="192.168.1.107" user="triton" password="triton" timeout="5.0" env-loader="/home/triton/env.sh" />
</launch>
```
- Then we run the source ros_ws/devel/setup.bash
- Then we run roslaunch crazyswarm hover.launch
- There are no errors, and Rviz shows the crazy flies 
- Pressing the start button on the controller shows take off but crazyflie(s) do not do anything. 

------------------------
running chooser.py works and all options work. Battery shows battery, reboot reboots. 
-We have tried updating the firmware by run the flash options many times from chooser, as well as using cload. No errors

------------------------
We have tried the set up process above through running roslaunch crazyswarm hover.launch, opened a new terminal, and run python hello_world.py

-It runs and gives no output
-it has no errors
-the process doesn't end without control c.
-------------------------------------------------------------------------
We are using Optitrack as the motion capture type in the hover.launch file
-we have both local and server IP's set 
---------------------------------
Would the fact that these are crazyflie 2.1 instead of 2.0 make a difference? 
-Note: the hello_world.py --sim works without fault. 
Any ideas on a solution? 
Thank you.

**Comment by jpreiss** [comment]
If `hello_world.py` does not terminate, try putting some print statements in between the lines to see how far it gets. Most likely, there is some problem in communication between the script and the Crazyswarm server, and it is getting stuck waiting for ROS services in the `Crazyswarm()` constructor. Make sure you source `devel/setup.bash` in the terminal from which you run `hello_world.py`.

However, if you see the `Takeoff!` printout from the Crazyswarm server when pressing the start button on the controller, but nothing happens, that suggests a problem in communication between the Crazyswarm server and the Crazyflies.

**Reply by elim17** [reply]
Thank you. Could you tell me what I should try to fixed the connection issue between crazy swarm and crazy flies? The cfclient works to fly the crazy flies manually, and the chooser seems to be able to connect with the crazy flies to flash, get battery, reboot, turn off ect. Thank you.

**Reply by whoenig** [reply]
Perhaps the radio firmware wasn't updated? Sounds like the broadcasts don't work, while unicast communication is operational.

**Reply by elim17** [reply]
Ok I will try that when I am back in the lab. Thank you.

**Reply by elim17** [reply]
I was able to flash the crazy radio based on your documentation here [Option 2: Use the prebuilt binary:

python crazyradio-firmware/usbtools/launchBootloader.py
sudo python crazyradio-firmware/usbtools/nrfbootload.py flash prebuilt/cradio.bin
Now unplug and re-plug the radio. You can check the version using rosrun crazyflie_tools scan -v, which should report Found Crazyradio with version 99.55.](url) The exact instructions I followed were: 
```
Option 2: Use the prebuilt binary:

1. python crazyradio-firmware/usbtools/launchBootloader.py
2. sudo python crazyradio-firmware/usbtools/nrfbootload.py flash prebuilt/cradio.bin
3. Now unplug and re-plug the radio. You can check the version using rosrun crazyflie_tools scan -v, which should report Found Crazyradio with version 99.55.
```

I can still use all functionality in the chooser.py however I was not able to test if this fixed anything because I'm waiting for this [discussion()](https://github.com/USC-ACTLab/crazyswarm/discussions/332) to pan out. It appears that we have issues connecting to the motion capture system in the correct manner. Will update when other issue is resolved and this issue can be tested.

**Reply by elim17** [reply]
We were able to get our connection issues with Motive and NatNet figured out [https://github.com/USC-ACTLab/crazyswarm/discussions/332](url). Solving the connection issues also solved the issues here as well and the drones were able to take off and hello_world.py also worked! Thank you for the help!

**Reply by jpreiss** [reply]
glad to hear it!

**Reply by OP-Follower** [reply]
> If `hello_world.py` does not terminate, try putting some print statements in between the lines to see how far it gets. Most likely, there is some problem in communication between the script and the Crazyswarm server, and it is getting stuck waiting for ROS services in the `Crazyswarm()` constructor. Make sure you source `devel/setup.bash` in the terminal from which you run `hello_world.py`.
> 
> However, if you see the `Takeoff!` printout from the Crazyswarm server when pressing the start button on the controller, but nothing happens, that suggests a problem in communication between the Crazyswarm server and the Crazyflies.

Sorry i have the problem in communication between the script and the Crazyswarm server. I made sure had sourced the "setup.bash", but it did not work. I can use joystick to make the cf "takeoff" and "land" after roslaunch "hover_swarm" and "mocap_helper", but when use the script it got stucked.

Hope you can give me some solutions or suggestions. Thanks in advance!!!
